Zend/zend_compile.c

```diff
@@ -6396,19 +6398,18 @@ void zend_compile_prop_decl(zend_ast *ast, zend_ast *type_ast, uint32_t flags) /
 					"Property %s::$%s cannot have type %s",
 					ZSTR_VAL(ce->name), ZSTR_VAL(name), ZSTR_VAL(str));
 			}
+		} else if (flags & ZEND_ACC_FINAL) {
+			zend_error_noreturn(E_COMPILE_ERROR,
```

Comment:
I missed this part of the RFC. https://wiki.php.net/rfc/write_once_properties#compile-type_restrictions

Sadly, there's no way to write `final public resource $fileHandle;`, because `resource` isn't a type declaration in PHP (to give the flexibility to switch resources to objects).

I'd wonder if it'd be possible to remove the syntactic restriction internally (for write-once properties) and represent `final` as a property type info with "every type" if there is no explicit type, i.e. `null|bool|int|float|string|object|array|<resource>` (MAY_BE_RESOURCE) - I don't know if that's actually easy to do, though.

- The alternative I have has its own drawbacks of implicitly and unexpectedly using the uninitialized state of typed properties, so enforcing manually adding a type is probably the best approach for now

Reply:
I think something similar has already been discussed on the mailing list before. That is, implicitly treating all non-typed properties to be `mixed`. I think it would be much simpler and more straightforward to just wait for the `mixed` type instead. (I really cheer for it to reach PHP 8).